# Exploring the Hacker News Project

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

You can find the data set at https://www.kaggle.com/hacker-news/hacker-news-posts

## This is what we will accomplish in this guided project:

    Setting a goal for the project.
    Collecting and sorting the data.
    Reformatting and cleaning the data to prepare it for analysis.
    Analyzing the data

Below are descriptions of the columns:

    id: The unique identifier from Hacker News for the post
    title: The title of the post
    url: The URL that the posts links to, if it the post has a URL
    num_points: The number of points the post acquired, calculated     as the total number of upvotes minus the total number of downvotes
    num_comments: The number of comments that were made on the post
    author: The username of the person who submitted the post
    created_at: The date and time at which the post was submitted
    
We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. Below are a couple examples:

Ask HN: How to improve my personal website?

Ask HN: Am I the only one outraged by Twitter shutting down share counts?

Ask HN: Aby recent changes to CSS that broke mobile?

Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. Below are a couple of examples:

Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'

Show HN: Something pointless I made

Show HN: Shanhu.io, a programming playground powered by e8vm

We'll compare these two types of posts to determine the following:

    Do Ask HN or Show HN receive more comments on average?
    Do posts created at a certain time receive more comments on average?

We will start by importing the libraries we need and reading the data set into a list of lists.




In [1]:
from csv import reader

# use the python built-in function open()
# to open the children.csv file
opened_file = open('hacker_news.csv')
reader = reader(opened_file)
hn = list(reader)

#hn = hn[:5]

print(hn[:5])



[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [2]:
headers = hn[:1]
print(headers)

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]


In [3]:
hn = hn[1:]
print(hn[:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


### Let's separate posts beginning with Ask HN and Show HN (and case variations) into two different lists .

In [4]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    #print(title)
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else : other_posts.append(row)
print(len(ask_posts))

1744


In [5]:
print(ask_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


In [6]:
print(len(show_posts))

1162


In [7]:
print(show_posts[:5])

[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05'], ['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11'], ['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']]


In [8]:
 print(len(other_posts))

17194


In [9]:
print(other_posts[:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


### Next, let's determine if ask posts or show posts receive more comments on average.

In [10]:
total_ask_comments = 0
for post in ask_posts:
    num_comments = int(post[4])
    total_ask_comments = total_ask_comments + num_comments
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

    

14.038417431192661


In [11]:
total_show_comments = 0
for post in show_posts:
    num_comments = int(post[4])
    total_show_comments = total_show_comments + num_comments
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


Ask posts receive more comments on average. This is because ask posts has 14.0 average comments yet show posts have 10.3 average comments

In [12]:
import datetime as dt
result_list = []
for post in ask_posts:
    created_at = post[6]
    num_comments = int(post[4])
    xx = [created_at, num_comments]
    result_list.append(xx)
    
print(result_list[:4])

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3]]


In [13]:
counts_by_hour = {}
comments_by_hour = {}
for row in result_list:
    created_at = dt.datetime.strptime(row[0], '%m/%d/%Y %H:%M')
    created_at_hour = created_at.strftime('%H')
    if created_at_hour in counts_by_hour:
        counts_by_hour[created_at_hour] += 1
        comments_by_hour[created_at_hour] += row[1]
    else:
        counts_by_hour[created_at_hour] = 1
        comments_by_hour[created_at_hour] = row[1]
        
print(counts_by_hour)
print(comments_by_hour)

{'18': 109, '16': 108, '05': 46, '02': 58, '20': 80, '21': 109, '04': 47, '15': 116, '10': 59, '09': 45, '03': 54, '07': 34, '01': 60, '14': 107, '13': 85, '11': 58, '23': 68, '19': 110, '08': 48, '00': 55, '17': 100, '06': 44, '22': 71, '12': 73}
{'18': 1439, '16': 1814, '05': 464, '02': 1381, '20': 1722, '21': 1745, '04': 337, '15': 4477, '10': 793, '09': 251, '03': 421, '07': 267, '01': 683, '14': 1416, '13': 1253, '11': 641, '23': 543, '19': 1188, '08': 492, '00': 447, '17': 1146, '06': 397, '22': 479, '12': 687}


### Next, we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day. 

In [15]:
avg_by_hour = []

for hour, counts in counts_by_hour.items():
    avg_comments = comments_by_hour[hour] / counts
    avg_by_hour.append([hour, avg_comments])
    
print(avg_by_hour)



[['18', 13.20183486238532], ['16', 16.796296296296298], ['05', 10.08695652173913], ['02', 23.810344827586206], ['20', 21.525], ['21', 16.009174311926607], ['04', 7.170212765957447], ['15', 38.5948275862069], ['10', 13.440677966101696], ['09', 5.5777777777777775], ['03', 7.796296296296297], ['07', 7.852941176470588], ['01', 11.383333333333333], ['14', 13.233644859813085], ['13', 14.741176470588234], ['11', 11.051724137931034], ['23', 7.985294117647059], ['19', 10.8], ['08', 10.25], ['00', 8.127272727272727], ['17', 11.46], ['06', 9.022727272727273], ['22', 6.746478873239437], ['12', 9.41095890410959]]


### Although we now have the results we need, this format makes it hard to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [16]:
# Sort the average comments list

swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
swap_avg_by_hour

[[13.20183486238532, '18'],
 [16.796296296296298, '16'],
 [10.08695652173913, '05'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.009174311926607, '21'],
 [7.170212765957447, '04'],
 [38.5948275862069, '15'],
 [13.440677966101696, '10'],
 [5.5777777777777775, '09'],
 [7.796296296296297, '03'],
 [7.852941176470588, '07'],
 [11.383333333333333, '01'],
 [13.233644859813085, '14'],
 [14.741176470588234, '13'],
 [11.051724137931034, '11'],
 [7.985294117647059, '23'],
 [10.8, '19'],
 [10.25, '08'],
 [8.127272727272727, '00'],
 [11.46, '17'],
 [9.022727272727273, '06'],
 [6.746478873239437, '22'],
 [9.41095890410959, '12']]

### Next we use the sorted() function to sort swap_avg_by_hour in descending order. 

In [18]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap[:5]

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21']]

In [19]:
template = '{hour}: {avg:.2f} average comments per post'

for row in sorted_swap:
    hour = dt.datetime.strptime(row[1], '%H').strftime('%H:00')
    output = template.format(hour=hour, avg=row[0])
    print(output)

15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post
13:00: 14.74 average comments per post
10:00: 13.44 average comments per post
14:00: 13.23 average comments per post
18:00: 13.20 average comments per post
17:00: 11.46 average comments per post
01:00: 11.38 average comments per post
11:00: 11.05 average comments per post
19:00: 10.80 average comments per post
08:00: 10.25 average comments per post
05:00: 10.09 average comments per post
12:00: 9.41 average comments per post
06:00: 9.02 average comments per post
00:00: 8.13 average comments per post
23:00: 7.99 average comments per post
07:00: 7.85 average comments per post
03:00: 7.80 average comments per post
04:00: 7.17 average comments per post
22:00: 6.75 average comments per post
09:00: 5.58 average comments per post


## Conclusion

In order to have a higher chance of receiving more comments on an Ask Post,a post should be created at 15:00 UTC
